docs/GHAFileExplained.md

```diff
@@ -0,0 +1,168 @@
+---
+layout: page
+title: GitHub Actions CI workflow file explained
+---
+
+Below is the
+[gha.dist.yml](https://github.com/moodlehq/moodle-plugin-ci/blob/master/gha.dist.yml)
+file with comments added to explain what each section is doing. For additional
+information please refer to [workflow syntax reference](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions).
+
+If you are familiar with Travis, this should be strightforward to understand
+the syntax, you may find this [migration
+manual](https://docs.github.com/en/actions/learn-github-actions/migrating-from-travis-ci-to-github-actions)
+useful.
+
+```yaml
+# Title of the workflow
+name: Moodle Plugin CI
+
+# Run this workflow every time a new commit pushed to your repository or PR
+# created.
+on: [push, pull_request]
+
+jobs:
+  # Set the job key. The key is displayed as the job name
+  # when a job name is not provided
+  test:
+    # Virtual environment to use.
+    runs-on: ubuntu-18.04
+
+    # DB services you need for testing.
+    services:
+      postgres:
+        image: postgres:9.6
+        env:
+          POSTGRES_USER: 'postgres'
+          POSTGRES_HOST_AUTH_METHOD: 'trust'
+        ports:
+          - 5432:5432
+        options: --health-cmd pg_isready --health-interval 10s --health-timeout 5s --health-retries 3
+      mariadb:
+        image: mariadb:10
+        env:
+          MYSQL_USER: 'root'
+          MYSQL_ALLOW_EMPTY_PASSWORD: "true"
+        ports:
+          - 3306:3306
+        options: --health-cmd="mysqladmin ping" --health-interval 10s --health-timeout 5s --health-retries 3
+
+    # Determines build matrix. This is a list of PHP versions, databases and
+    # branches to test our project against. For each combination a separate
+    # build will be created. For example below 6 builds will be created in
+    # total (7.2-pgsql, 7.2-mariadb, 7.3-pgsql, 7.3-mariadb, etc.). If we add
+    # another branch, total number of builds will become 12.
+    strategy:
+      fail-fast: false
+      matrix:
+        php: ['7.2', '7.3', '7.4']
+        moodle-branch: ['MOODLE_310_STABLE']
+        database: [pgsql, mariadb]
+
+    # There is an alterantive way allowing to define explicitly define which php, moodle-branch
+    # and database to use:
+    # matrix:
+    #   include:
+    #     - php: '7.4'
+    #       moodle-branch: 'MOODLE_310_STABLE'
+    #       database: pgsql
+    #     - php: '7.3'
+    #       moodle-branch: 'MOODLE_310_STABLE'
+    #       database: mariadb
+    #     - php: '7.2'
+    #       moodle-branch: 'MOODLE_39_STABLE'
+    #       database: pgsql
+
+    steps:
+      # Check out this repository code in ./plugin directory
+      - name: Check out repository code
+        uses: actions/checkout@v2
+        with:
+          path: plugin
+
+      - name: Install node
+        uses: actions/setup-node@v1
+        with:
+          node-version: '14.15.0'
```

Comment:
I am seeing this step in all existing examples. And yet I do not get why it is needed. moodle-plugin-ci should install node version specified in .nvmrc itself, no? At the moment it does not work due to permissions but that seems addressed by @timhunt in https://github.com/moodlehq/moodle-plugin-ci/commit/a03897b789d711288f6ba7248fb987911e9624c7

Surely I am just missing something. TIA for clarification.

Reply:
BTW, did someone play with something like https://github.com/actions/cache/blob/main/examples.md#node---npm already?

Reply:
Yes, it probably got copied from early trials. `nvm` is not there, so `moodle-plugin-ci` can't install it. In fact, thanks for highlighting @mudrd8mz , I will test if https://github.com/dcodeIO/setup-node-nvm can help us.

Reply:
I did some investigation, nvm is in fact available in all ubuntu environments (https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu1804-README.md), so we can use it (https://github.com/moodlehq/moodle-plugin-ci/pull/73). Thanks @mudrd8mz!